mitiq/folding_cirq.py

```diff
@@ -1,10 +1,36 @@
 """Functions to fold gates in Cirq circuits."""
 from copy import deepcopy
-from typing import (Any, Callable, Iterable, List, Optional, Tuple)
+from typing import (Any, Callable, Iterable, List, Optional, Sequence, Tuple)
 
 import numpy as np
 
-from cirq import (Circuit, InsertStrategy, inverse)
+from cirq import (Circuit, InsertStrategy, inverse, ops)
+
+
+# Helper functions
+def _is_measurement(op: ops.Operation) -> bool:
+    """Returns true if the operation's gate is a measurement, else False.
+
+    Args:
+        op: Gate operation.
+    """
+    return isinstance(op.gate, ops.measurement_gate.MeasurementGate)
+
+
+def _pop_measurements(circuit: Circuit):
+    """Removes all measurements from a circuit."""
+    measurements_generator = circuit.findall_operations(_is_measurement)
+    # Make measurements a list of lists so moment indices can be updated after gates are folded
+    measurements = []
+    for m in measurements_generator:
+        measurements.append(list(m))
+    circuit.batch_remove(measurements)
+    return measurements
+
+
+def _add_measurements(circuit: Circuit, measurements: Iterable[Tuple[int, ops.Operation]]) -> None:
```

Comment:
Does this need to be a standalone function? Could `circuit.batch_insert(measurements) be used in place?

Reply:
My thinking was to abstract one layer away from Cirq in case they change anything. That way we can keep the usage of `_add_measurements` the same but change its implementation if necessary.